src/finality/grandpa/warp_sync.rs

```diff
@@ -0,0 +1,96 @@
+// Substrate-lite
+// Copyright (C) 2019-2020  Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::chain::chain_information::{ChainInformation, ChainInformationFinality};
+use crate::finality::justification::verify::{
+    verify, Config as VerifyConfig, Error as VerifyError,
+};
+use crate::header::{DigestItemRef, GrandpaConsensusLogRef};
+use crate::network::protocol::GrandpaWarpSyncResponseFragment;
+use std::convert::TryFrom;
+
+pub struct Verifier {
+    index: usize,
+    authorities_list: Vec<[u8; 32]>,
+    fragments: Vec<GrandpaWarpSyncResponseFragment>,
+}
+
+impl Verifier {
+    pub fn new(
+        genesis_chain_infomation: &ChainInformation,
+        warp_sync_response_fragments: Vec<GrandpaWarpSyncResponseFragment>,
+    ) -> Self {
+        let authorities_list = match &genesis_chain_infomation.finality {
+            ChainInformationFinality::Grandpa {
+                finalized_triggered_authorities,
+                ..
+            } => finalized_triggered_authorities
+                .iter()
+                .map(|auth| auth.public_key)
+                .collect(),
+            _ => unimplemented!(),
```

Comment:
This should return an error rather than panicking, or at least add a `// TODO:` saying so.